extensions/sql-migration/src/models/loginMigrationModel.ts

```diff
@@ -0,0 +1,247 @@
+	Pending = 'Pending',
+	Running = 'Running',
+	Succeeded = 'Succeeded',
+	Failed = 'Failed',
+	Canceled = 'Canceled',
+}
+
+export const LoginMigrationStateToMultiStepState: constants.LookupTable<MultiStepState> = {
+	[LoginMigrationState.Pending]: MultiStepState.Pending,
+	[LoginMigrationState.Running]: MultiStepState.Running,
+	[LoginMigrationState.Succeeded]: MultiStepState.Succeeded,
+	[LoginMigrationState.Failed]: MultiStepState.Failed,
+	[LoginMigrationState.Canceled]: MultiStepState.Canceled,
+	default: MultiStepState.Pending
+};
+
+
+export interface LoginMigrationStepState {
+	loginName: string;
+	stepName: LoginMigrationStep;
+	status: LoginMigrationState;
+	errors: string[];
+}
+
+export interface Login {
+	loginName: string;
+	overallStatus: LoginMigrationState;
+	statusPerStep: Map<LoginMigrationStep, LoginMigrationStepState>;
+}
+
+export class LoginMigrationModel {
+	public resultsPerStep: Map<mssql.LoginMigrationStep, mssql.StartLoginMigrationResult>;
+	public collectedSourceLogins: boolean = false;
+	public collectedTargetLogins: boolean = false;;
+	public loginsOnSource: LoginTableInfo[] = [];
+	public loginsOnTarget: string[] = [];
+	private _currentStepIdx: number = 0;;
+	private _logins: Map<string, Login>;
+	private _loginMigrationSteps: LoginMigrationStep[] = [];
+
+	constructor() {
+		this.resultsPerStep = new Map<mssql.LoginMigrationStep, mssql.StartLoginMigrationResult>();
+		this._logins = new Map<string, Login>();
+		this.SetLoginMigrationSteps();
+	}
+
+	public get currentStep(): LoginMigrationStep {
+		return this._currentStepIdx >= this._loginMigrationSteps.length ? LoginMigrationStep.MigrationCompleted : this._loginMigrationSteps[this._currentStepIdx];
+	}
+
+	public get isMigrationComplete(): boolean {
+		return this._currentStepIdx === this._loginMigrationSteps.length;
+	}
+
+	public AddLoginMigrationResults(step: LoginMigrationStep, newResult: mssql.StartLoginMigrationResult): void {
+		const exceptionMap = this._getExceptionMapWithNormalizedKeys(newResult.exceptionMap);
+		this._currentStepIdx = this._loginMigrationSteps.findIndex(s => s === step) + 1;
+
+		for (const loginName of this._logins.keys()) {
+			const status = loginName in exceptionMap ? LoginMigrationState.Failed : LoginMigrationState.Succeeded;
+			let errors = loginName in exceptionMap ? this._extractErrors(exceptionMap, loginName) : [];
+			this._addStepStateForLogin(loginName, step, status, errors);
+
+			if (this.isMigrationComplete) {
+				const loginStatus = this._didAnyStepFail(loginName) ? LoginMigrationState.Failed : LoginMigrationState.Succeeded;
+				this._markLoginStatus(loginName, loginStatus);
+			}
+		}
+	}
+
+	public ReportException(step: LoginMigrationStep, error: any): void {
+		this._currentStepIdx = this._loginMigrationSteps.findIndex(s => s === step) + 1;
+
+		for (const loginName of this._logins.keys()) {
+			// Mark current step as failed with the error message and mark remaining messages as canceled
```

Comment:
does const LoginName work here or should it be let LoginName so it can change while iterating

Reply:
const loginName works here. It's the key that is used to update the state of all the steps and shouldn't change